File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/mobs/guts/YogsTeeth.java
package com.nyrds.pixeldungeon.mobs.guts;

import com.nyrds.pixeldungeon.effects.Devour;
import com.nyrds.pixeldungeon.mechanics.NamedEntityKind;
import com.nyrds.platform.audio.Sample;
import com.watabou.pixeldungeon.Assets;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.blobs.ToxicGas;
import com.watabou.pixeldungeon.actors.buffs.Amok;
import com.watabou.pixeldungeon.actors.buffs.Bleeding;
import com.watabou.pixeldungeon.actors.buffs.Buff;
import com.watabou.pixeldungeon.actors.buffs.Burning;
import com.watabou.pixeldungeon.actors.buffs.Paralysis;
import com.watabou.pixeldungeon.actors.buffs.Sleep;
import com.watabou.pixeldungeon.actors.buffs.Terror;
import com.watabou.pixeldungeon.actors.mobs.Mob;
import com.watabou.utils.Random;

import org.jetbrains.annotations.NotNull;

/**
 * Created by DeadDie on 12.02.2016
 */
public class YogsTeeth extends Mob {
    {
        hp(ht(350));
        baseDefenseSkill = 44;
        baseAttackSkill  = 46;
        dmgMin = 50;
        dmgMax = 80;
        dr = 21;

        expForKill = 26;

        addResistance(ToxicGas.class);

        addImmunity(Paralysis.class);
        addImmunity(Amok.class);
        addImmunity(Sleep.class);
        addImmunity(Terror.class);
        addImmunity(Burning.class);
    }

    @Override
    public void damage(int dmg, @NotNull NamedEntityKind src) {
        for (Mob mob : level().mobs) {
            mob.beckon(getPos());
        }

        super.damage(dmg, src);
    }

    @Override
    public int attackProc(@NotNull Char enemy, int damage ) {
        //Life drain proc
        if (Random.Int(3) == 1){
            heal(damage, this);
        }

        //Bleeding proc
        if (Random.Int(3) == 1){
            Buff.affect(enemy, Bleeding.class).level(damage);
        }

        //Double damage proc
        if (Random.Int(3) == 1){
            Devour.hit(enemy);
            Sample.INSTANCE.play(Assets.SND_BITE);
            return damage*2;
        }
        return damage;
    }

    @Override
    public boolean canBePet() {
        return false;
    }
}

File: scripts/mobs/YogsTeeth.lua
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

-- java YogsTeeth: being hurt draws every mob to it; each hit rolls three
-- independent procs - life drain, bleeding as strong as the hit, and a
-- devoured (double) hit.
local function beckonAll(self)
    local mobs = RPD.Dungeon.level:getMobs()
    for i = 1, #mobs do
        mobs[i]:beckon(self:getPos())
    end
end

return mob.init{
    damage = function(self, dmg, src)
        beckonAll(self)
    end,
    attackProc = function(self, enemy, dmg)
        if enemy ~= nil then
            if math.random(3) == 1 then
                self:heal(dmg, self)
            end
            if math.random(3) == 1 then
                local bleeding = RPD.Buffs.Buff:affect(enemy, "Bleeding")
                bleeding:level(dmg)
            end
            if math.random(3) == 1 then
                RPD.Devour:hit(enemy)
                RPD.playSound("snd_bite")
                return dmg * 2
            end
        end
        return dmg
    end
}
